README.md
# OSDOOBEUI

An Out-of-the-Box Experience like user interface for Operating System Deployments

## Description

A PowerShell Driven UI that looks and feels like Windows 10 OOBE. Used in SCCM/MDT TaskSequence Bare-metal deployments

## Requirements

To support the UI in windows PE, these features must be installed:

- DISM Cmdlets
- Microsoft Data Access Component (MDAC/ADO) support
- Windows PowerShell

The recommended command to run the UI is:

#### **OSDOOBEUI.ps1**

Work-In-Progress
The UI is designed o look and feel like Windows 10 OOBE startup wizard but with more control

```bash
"%DEPLOYROOT%\tools\%architecture%\ServiceUI.exe" -process:TSProgressUI.exe "%SYSTEMROOT%\System32\WindowsPowerShell\v1.0\powershell.exe" -ExecutionPolicy Bypass -Sta -File "%SCRIPTROOT%\Custom\OSDOOBEUI\OSDOOBEUI.ps1"
```


#### **OSDOOBEUI_SinglePage.ps1**
 The UI is designed to be a single page with the option to add an additional app page if needed. The UI will look like this:
 ![UI](UI_Screenshots/OOBEUIWPF_Blank-singlepage.png)
 
 ![Validate](UI_Screenshots/OOBEUIWPF_Validate-singlepage.png)

```bash
"%DEPLOYROOT%\tools\%architecture%\ServiceUI.exe" -process:TSProgressUI.exe "%SYSTEMROOT%\System32\WindowsPowerShell\v1.0\powershell.exe" -ExecutionPolicy Bypass -Sta -File "%SCRIPTROOT%\Custom\OSDOOBEUI\OSDOOBEUI_SinglePage.ps1" -ConfigPath OOBEUIWPF.config
```

### Current Features

- **Splash screen**: can be enabled or disabled by the config file. The splash screen is useful when hiding the PowerShell window screen while the UI is loading and task sequence progress bar
![Splash screen](UI_Screenshots/splashscreen.PNG)

## Configurable Items

### Options Configuration

All settings are configurable within the **OSDOOBEUI.config** file

- **Logo1Position _[string]_**;   Changes the position of the logo in menu. Options: Left, Right, Both, Hidden; default is Left.
- **Logo2Position _[string]_**;   Changes the position of the logo in menu. Options: Left, Right, Both, Hidden; default is right
- **Logo1File _[string]_**;       Location where log file exists. (Recommend size is 100x100)
- **Logo2File _[string]_**;       Location where log file exists. (Recommend size is 100x100)
- **FormVariable _[string]_**;   Set the variable used for the Object sin XAML. No need to change
- **VerboseMode _[Boolean]_**;   Output verbose messages to PowerShell window
- **DebugMode _[Boolean]_**;     Output additional messages to PowerShell window
- **Test mode _[Boolean]_**;     Does not show the menu, just runs all functions and menu items
After menu is played/started, run these commands in PowerShell to trigger validation tests:

```powershell
        $OSDOOBEUI_txtComputerName.text = 'ADPROD12345PAW'
        $ComputerNameObject = $OSDOOBEUI_txtComputerName
        Validate-ComputerNameRules $ComputerNameObject
```

- **BackgroundColor _[string]_**;Changes the main background color of the UI
    **NOTE:** Buttons are not changeable. Recommend keeping background a blue tone

### Page Options

- **[Still Testing] SinglePageOnly _[Boolean]_**;  Display the UI as a single page. App page can be enabled if set to true
- **ShowAppSelection _[Boolean]_**; Display the App page (used for both single and multi page UI's)
![Apps](UI_Screenshots/OOBEUIWPF_Apps.png)
### Control Configuration

- **[Still Testing] AutoGenerateName _[string]_**; Auto generates name based option selected _[AD, SQL or Locale, disabled]_. Using a SQL table and AD, name will be autogenerated and validated, Locale uses local config from network locale
- **[Still Testing] GenerateNameMethod _[string]_**; Auto generates name based option selected _[AD, SQL,ODJFile,ODJBlob,Locale,TSEnv,Clear].

- **[Still Testing] GenerateNameSource _[string]_**; Used only with ODJ set in GenerateNameMethod Property; specifies location to pull ODJ file. Path can be a local, network share, or a URL (Will use invoke method). ODJ file path will search for odj files that follow the naming format. Must be named: <assettag>_<serialnumber>_<computername>.odj. Using a SQL table and AD, name will be autogenerated and validated, Locale uses local config from network locale
  NOTE: ODJ set custom properties: ODJ_FilePath,ODJ_BlobData
- **OverWriteUIControlByTS _[Boolean]_**; Determines if tasksequence controls the UI vs Config (only changes UI_Control section)
                                                        TS variables needs to mirror control but with UI prefix added (eg. <Control_ShowSplashScreen>True</Control_ShowSplashScreen> -> UIControl_ShowSplashScreen=True)
- **ShowSplashScreen _[Boolean]_**;       Display splash screen prior to menu loaded. Hides verbose command window
- **ShowSiteCode _[Boolean]_**;           Displays site code. Site code can be changed is displayed
- **ShowSiteListSelection _[Boolean]_**;  Dropdown of all sites with site ID and site Code(if exists)
- **EnableNetworkDetection _[Boolean]_**; Select site locales based on current IP
- **ValidateNameRules _[Boolean]_**;      Uses Generation Rule Sets to ensure name is valid. Does not check if name is available on domain
- **ShowClassificationProperty _[string]_**; Displays classification property in Identity field with appropriate color based on device name (Values to choose from: Id,Level,Type,None)
- **ShowDomainOUListSelection _[Boolean]_**; Dropdown to select the OU to join to. Can be controlled by site locale
- **AllowCustomDomain _[Boolean]_**;      If enabled the domain field is a fillable textbox, if disabled the domain field is a dropdown (populated by Locale Domain Info)
- **AllowWorkgroupJoin _[Boolean]_**;     If set to false, the workgroup option is filtered out.
- **AllowSiteSelection _[Boolean]_**;     Enables the ability to change the Site ID from selection list (Must have ShowSiteListSelection enabled)
- **FilterAccountDomainType _[string]_**; Set this value to match a Locale Domain Type property. If set, the matching domain by classification will be filled in for the account domain
- **FilterDomainProperty _[string]_**;    Value is used when matching domain property to classification id property. Currently supports only classification id filter.
- **AllowRuleBypassModeKey _[string]_**;  If set, the validation rules can be bypassed by pressing shift before clicking validate. This allows custom names. Other validations are still checked
- **HideDomainList _[Boolean]_**;         If this is set to true, the domains fields will be hidden and no validation or variables will be set; overwrites ShowDomainOUListSelection
- **HideDomainCreds _[Boolean]_**;        If this is set to true, the credential fields will be hidden and no validation or variables will be set
    -->

AllowRuleBypassModeKey screenshot
![Apps](UI_Screenshots/OOBEUIWPF_Bypassmode.png)

### Generation Rules

This is highly configurable and can have up to 5 areas. Please read the [config](OSDOOBEUI.config) file for more details

The varname can be renamed. however keep the **ID** the same

- **IDMachineType _[table]_**; Value is controlled based computer name field (eg._**;A**DPROP12345PAW"_).

example:
```xml
<rulesets Id="Id1" Name="Machine Type" VarName="IDMachineType" MustExist="True" >
    <rule Char="A" Name="Administrative"                VarValue="Admin" />
    <rule Char="D" Name="Developer"                     VarValue="Developer" />
    <rule Char="R" Name="Training"                      VarValue="Training" />
</rulesets>
```

- **IDFormFactor _[table]_**; Value is controlled based computer name field (eg. _A**D**PROP12345PAW"_).

example:
```xml
<rulesets Id="Id2" Name="Form Factor" VarName="IDFormFactor" MustExist="True" >
    <rule Char="D" Name="Desktop"                       VarValue="Desktop" />
    <rule Char="H" Name="Virtual Desktop (VDI)"         VarValue="VDI" />
    <rule Char="L" Name="Laptop"                        VarValue="Laptop" />
    <rule Char="U" Name="Virtual Machine"               VarValue="VM" />
</rulesets>
```

- **SiteID _[stable]_**; Value is controlled based computer name field (eg. _AD**PROP**12345PAW"_).

example:
```xml
<rulesets Id="Id3" Name="Site Identifier" VarName="SiteID" MustExist="True" >
    <rule Name="Local Sites" GetVariable="MenuLocaleSiteList" MatchProperty="ID" DisplayProperty="BaseLocation" SetVariable="SiteID" />
</rulesets>
```

- **IDMachineClass _[table]_**; Value is controlled based computer name field (eg. _ALDTLAB**12**3456PA_). Value is determined by the first and/or second digits.

example:
```xml
<rulesets Id="Id4" Name="Machine Class" VarName="IDMachineClass" MinCharIdentifier="5" MustExist="True" >
    <rule Char="0#####" Name="Corporation"          VarValue="Corp" />
    <rule Char="1#####" Name="Laboratory"             VarValue="Lab" />
</rulesets>
```

- **IDMachineRole _[table]_**; Value is controlled based computer name field (eg. _ALDTLAB123456**PA**_). Value can be Null. Variable name can be changed in ruleset's _VarName_ property

example:
```xml
<rulesets Id="Id5" Name="Role Identifier" VarName="IDMachineRole" MustExist="True" >
    <rule Char="dPAW" Name="Privilege Access Workstation"             VarValue="PAW" />
    <rule Char="SSW" Name="Security Access Workstation"              VarValue="SAW" />
    <rule Char="JMP" Name="Jump Box Workstation"                     VarValue="JMP" />
</rulesets>
```
### Classification list

If enabled, list is pre selected based on Computer Machine AS Rule; Searches for Type

example:
```xml
<Locale_Classifications>
	<classification Id="Prod" Level="Confidential"	Type="Confidential"	Color="Green"  />
	<classification Id="Lab" Level="Laboratory"    Type="Laboratory"	Color="Yellow"	/>
</Locale_Classifications>
```
### Domain List

Value is controlled based on dropdown field or by computer name value and when Validate is pressed. Not all options are present:

example:
```xml
<Locale_Domains>
	<domain Name="LABD" FQDN="lab.contoso.com"       ClassId="Lab" Type="Join" />
	<domain Name="PROD" FQDN="prod.contoso.com"       ClassId="Prod" Type="Join" />
	<domain Name="PROD" FQDN="prod.contoso.com"       ClassId="Prod" Type="Auth" />
</Locale_Domains>
```
### DomainOU List

If enabled, this list will poplulate a dropdown based on site selected

example:
```xml
<Locale_DomainOUs>
		<OU Name="Lab" Domain="LABD" LDAPOU="OU='Test',OU='lab',DC='contoso',DC=com"    ClassId="Lab" />
		<OU Name="Production" Domain="PROD" LDAPOU="OU='Systems',OU='Prod',DC='contoso',DC=com"    ClassId="Prod" />
</Locale_DomainOUs>
```

### Site List

Value is controlled based on dropdown field or by computer name value and when Validate is pressed. Not all options are present:

example:
```xml
<Locale_Sites DisplayFormat="&lt;id&gt; - &lt;Baselocation&gt;" SiteCodeFormat=" _[&lt;SiteCode&gt;]" >
    <site ID="LAB" BaseLocation="Laboratory Site, USA"  TZ="MST" Region="US" SiteCode="LAB" Domain="LABD" />
    <site ID="PRO" BaseLocation="Production Site, USA"  TZ="CST" Region="US" SiteCode="PRO" Domain="PROD" />
</Locale_Sites>
```

### NetworkDetection List

If enabled, it will match the IP subnet with the Site code and preselect the option in the UI

example:
```xml
<Locale_NetworkDetection>
    <Network SiteId="TEST" CidrAddr="192.168.1.0/24" />
    <Network SiteId="SITEB" CidrAddr="10.11.0.0/24" />
    <Network SiteId="SITEA" CidrAddr="10.21.0.0/24" />
</Locale_NetworkDetection>
```

### Applications

```xml
<Menu_AppButtons>
	<item id="1" Name="Update Microsoft Office"      TSvar="MSO13" DefaultEnabled="No" Desc='This will attempt to install updates for office if detected. This can prolong the deployment' />
    <item id="2" Name="Install Microsoft InfoPath"   TSvar="MSOIP" DefaultEnabled="Yes" Desc='InfoPath 2013 is a forms-creation and data-gathering tool that helps organizations streamline business processes' />
    <item id="3" Name="Install Adobe Reader DC"      TSvar="ADRDC" DefaultEnabled="No"  Desc='Install Adobes PDF reader instead of using Windows built-in PDF reader' />
	<item id="4" Name="Install Mozilla Firefox"      TSvar="MOZF" DefaultEnabled="No" Desc='A browsers that supports specific DoD websites' />
	<item id="5" Name="Install Google Chrome"        TSvar="GCE" DefaultEnabled="No" Desc='A browsers that supports specific DoD websites' />
	<item id="7" Name="Run GPO Script"               TSvar="SCRIPT1" DefaultEnabled="No" />
	<item id="8" Name="Add CMtrace"                  TSvar="SCRIPT2" DefaultEnabled="Yes" Desc='A tool that was created by Microsoft to view their "*.log" live and works in different environments.'/>
</Menu_AppButtons>
```

## OSD variables

If a task sequence is detected, the output will be a TS variable

- **OSDComputerName _[string]_**; value is what is in field
- **OSDNetworkJoinType _[0]_**; set only if domain name field is a string other than _workgroup_
- **OSDJoinType _[workgroup, domain]_**; set to _domain_ if domain field is a string other than _workgroup_
- **OSDDomainName _[string]_**; value is what is in field
- **OSDDomainOUName _[LDAP string]_**;
- **OSDJoinDomainOUName _[LDAP string]_**;
- **OSDJoinAccount _[string]_**;
- **OSDJoinDomainName _[string]_**;
- **OSDJoinPassword _[string]_**;
- **CMSiteCode _[string]_**; if enabled in configuration the site code is a 3 character string
- **TimeZone _[[integer]_**;
- **OSDTimeZone _[time zone]_**;
- **TimeZoneName _[time zone]_**;
- **OSDWorkgroupName _[workgroup]_**; only available when typing in _workgroup_ in the domain field
- **OSDJoinWorkgroupName _[workgroup]_**; only available when typing in _workgroup_ in the domain field
- **OSDLocalAdminPassword _[string]_**;
- **TimeZoneName _[string]_**;

Here are some screenshots or errors presented in UI

# Additional Screenshots



![Error1](UI_Screenshots/OOBEUIWPF_Validate_error1-singlepage.png)

![Error2](UI_Screenshots/OOBEUIWPF_Validate_error2-singlepage.png)

CHANGELOG.md
# Change log for OOBEUIWPF.ps1

## 2.1.4 - Mar 18 2021

- Renamed script to OSDOOBEUI.ps1 to be more consistent with repo. 
- Changed AddButtons to AppButtons in XML; mistyped

## 2.1.3 - Mar 16 2021

- Changed Serial number to pull from WMI win32_BIOS; works with multiple platforms. 
- Removed task sequence variable checks for device info; always uses WMI
- Check computer name when initiated; uses TS variable _SMSTSMachineName
- Fixed Testmode; changed UI help cmdlet

## 2.1.2 - Mar 10, 2021

- Used Timezone spreadsheet instead of system timezones; accurately control index and time.
- Fixed domain and timezone pre selection when validating device name

## 2.1.1 - Mar 04, 2021

- Changed domainname to domainFQDN in Set-OSDDomainVariables function
- Added custom osd variable for domain join: domainname. Allows multiple 'Apply Network Settings' step in one TS using custom variable detection

## 2.1.0 - Dec 18, 2020

- Changed OSD values to log only when set.
- Changed parameters used when Control_HideDomainCreds and Control_HideDomainList are set.

## 2.0.9 - Dec 17, 2020

- Fixed process OSD variables when hiding fields; doing a check if value exist before setting it.

## 2.0.8 - Dec 16, 2020

- Added Control_HideDomainCreds and Control_HideDomainList; used manually for SCCM BareMetal Imaging and only want a computer name prompt
- Check for MININT in name during task sequence; set to computer name to null; otherwise set current name; used when enabling WinPE as Generate Name Method
- Changed Generate Name Method option WinPE to TSEnv; checks for OSDComputerName is not default name MININT;
- added clear option to Generate Name Method; nulls out any computer name
- Changed Set-OSDComputerName to Update-OSDComputerName

## 2.0.7 - Nov 03, 2020

- Added the ability to bypass validation using a key set in config; must be done at Begin step as well
- Added Function to convert (or flip) combobox to textbox; and visa versa
- Added outhost trigger to UI console in config in addition to logging; Outputs execution actions. must enable verbose/debug for additional output.
- Fixed domain account update process based on domain dropdown or txt field.

## 2.0.6 - Oct 20, 2020

- Removed MDT variables in the Set-OSDVariables.ps1; not used in OSD anymore. Causes issues with username and password

## 2.0.5 - Oct 18, 2020

- Fixed external site list path when importing; added function to check

## 2.0.4 - Oct 17, 2020

- Added System.Windows assembly; causes less crashing
- Fixed external site list path when importing; reverts to working path and not relative path
- Added two logo support for both UI and Splashscreen
- Fixed logging when using debug or verbose switch; only logs based on settings
- Added build version info to UI; only displays if in debug or verbose mode
- Added version control check based on changelog.md for more accurate version and date
- Added SYNOPSIS details to main script

## 2.0.3 - Oct 16, 2020

- Updated Splashscreen to display progress of menu status
- Moved Splashscreen functions to separate file; set Splashscreen runspace to a global scope
- Renamed Splashscreen functions to have 'UI' in the name; better consistency for all UI functions
- Fixed config path checks to look for rooted path, relative, and absolute
- Added Set-UIFieldElement to manage multiple parameter at once; shorting script for multiple element calls
- Add configuration display format control for site list in dropdown; set by xml file.

## 2.0.2 - Oct 14, 2020

- Added Application page control with function; identifies active begin and flips controls dynamically.
- Fixed workgroup option when not validating rules
- Disable computer name update during site list change when not validating rules
- Fixed Domain OU display control
- added TimeZonelist function for cleaner script

## 2.0.1 - Oct 12, 2020

- Updated logging for each ps1 file in Function folder
- Change ODJ to support blob (in unattend) or file option (from workgroup); changes the method of joining to domain
- Added MDT variables to domain join function (Set-OSDVariables.ps1)
- Added password generator as example and to password validation for ODJ join
- renamed to ZTIOOBEUI.ps1 to follow ZTI or LTI format.

## 1.9.9 - Oct 05, 2020

- Added logging to main script
- Added support to import an external site list; must be in CSV format.
- Fixed colors in UI message to output supported Foreground colors in PowerShell.
- Enforces TimeZone abbreviation converter to always use first selection
- Add the ability to display site list but have it disabled

## 1.9.8 - Sept 19, 2020

- Fixed ODJ UI to hide only if Blob content is there, if blank file exist, still prompt for creds
- Fixed focus on password and computer name
- Added the ability to overwrite config file using TS variables or parameter

## 1.9.7 - Sept 16, 2020

- Added Offline Domain Join feature. Hides Domain and username and password if found.
- Moved majority of functions to separate files

## 1.9.6 - Aug 19, 2020

- Changed OOBEWPFUI_OptionalPage.ps1 to OOBEWPFUI_SinglePage.ps1
- Fixed Update-ComputerNameLocale site locale changer
- fixed splash screen to close when menu opens; not after menu closes and not before menu opens. Added delay

## 1.9.5 - Aug 5, 2020

- Added example control in config.
- fixed hidden logo option during WPF process
- Update computer name change changer for site locale. dynamically grab site code

## 1.9.4 - July 20, 2020

- changed ShowClassificationColor to ShowClassificationProperty allowing to change the display text and color based on property in config.
- added classification variable output: Classification, ClassificationColor,ClassificationLevel,ClassificationType,ClassificationCaveat
- cleaned up debug/verbose output for easier reading
- excluded name change warning message on initial entry

## 1.9.3 - July 16, 2020

- add example text (light gray) within computer name and domain account fields

Functions/PlatformInfo.ps1

#region FUNCTION: convert chassis Types to friendly name
Function ConvertTo-ChassisType{
    [CmdletBinding()]
    Param($ChassisId)
    Switch ($ChassisId)
        {
            "1" {$Type = "Other"}
            "2" {$Type = "Virtual Machine"}
            "3" {$Type = "Desktop"}
            "4" {$type = "Low Profile Desktop"}
            "5" {$type = "Pizza Box"}
            "6" {$type = "Mini Tower"}
            "7" {$type = "Tower"}
            "8" {$type = "Portable"}
            "9" {$type = "Laptop"}
            "10" {$type = "Notebook"}
            "11" {$type = "Handheld"}
            "12" {$type = "Docking Station"}
            "13" {$type = "All-in-One"}
            "14" {$type = "Sub-Notebook"}
            "15" {$type = "Space Saving"}
            "16" {$type = "Lunch Box"}
            "17" {$type = "Main System Chassis"}
            "18" {$type = "Expansion Chassis"}
            "19" {$type = "Sub-Chassis"}
            "20" {$type = "Bus Expansion Chassis"}
            "21" {$type = "Peripheral Chassis"}
            "22" {$type = "Storage Chassis"}
            "23" {$type = "Rack Mount Chassis"}
            "24" {$type = "Sealed-Case PC"}
            Default {$type = "Unknown"}
         }
    Return $Type
}
#endregion

#region FUNCTION: Grab all machine platform details
Function Get-PlatformInfo {
# Returns device Manufacturer, Model and BIOS version, populating global variables for use in other functions/ validation
# Note that platformType is appended to psobject by Get-PlatformValid - type is manually defined by user to ensure accuracy
    [CmdletBinding()]
    [OutputType([PsObject])]
    Param()
    try{
        $CIMSystemEncloure = Get-CIMInstance Win32_SystemEnclosure -ErrorAction Stop
        $CIMComputerSystem = Get-CIMInstance CIM_ComputerSystem -ErrorAction Stop
        $CIMBios = Get-CIMInstance Win32_BIOS -ErrorAction Stop

        $ChassisType = ConvertTo-ChassisType -ChassisId $CIMSystemEncloure.chassistypes

        [boolean]$Is64Bit = [boolean]((Get-WmiObject -Class 'Win32_Processor' | Where-Object { $_.DeviceID -eq 'CPU0' } | Select-Object -ExpandProperty 'AddressWidth') -eq 64)
        If ($Is64Bit) { [string]$envOSArchitecture = '64-bit' } Else { [string]$envOSArchitecture = '32-bit' }

        New-Object -TypeName PsObject -Property @{
            "computerName" = [system.environment]::MachineName
            "computerDomain" = $CIMComputerSystem.Domain
            "platformBIOS" = $CIMBios.SMBIOSBIOSVersion
            "platformManufacturer" = $CIMComputerSystem.Manufacturer
            "platformModel" = $CIMComputerSystem.Model
            "AssetTag" = $CIMSystemEncloure.SMBiosAssetTag
            "SerialNumber" = $CIMBios.SerialNumber
            "Architecture" = $envOSArchitecture
            "Chassis" = $ChassisType
            }
    }
    catch{Write-Output "CRITICAL" "Failed to get information from Win32_Computersystem/ Win32_BIOS"}
}
#endregion